### PR TITLE
Tighten language on middlebox

### PR DESCRIPTION
**Pull request by richsalz**
Use "network intermediary" in all but once place.
Add forward/backward links to Appendix D
Add requirement to SHOULD use compatibility.

### CODE REVIEW

File: draft-ietf-tls-tls13.md [hunk 1]
```diff
@@ -3896,9 +3899,9 @@ TLS records are typed, which allows multiple higher-level protocols to
 be multiplexed over the same record layer. This document specifies
 four content types: handshake, application data, alert, and
 change_cipher_spec, with the latter being used only for
-compatibility purposes (see {{middlebox}}).
+compatibility purposes (see {{intermediary}}).
 
-An implementation may receive an unencrypted record of type
+{#dummy_css}An implementation may receive an unencrypted record of type
```
Comment (kaduk) [inline draft-ietf-tls-tls13.md]: Are we talking cascading style sheets or change cipher spec?

File: draft-ietf-tls-tls13.md [hunk 2]
```diff
@@ -5560,17 +5563,21 @@ ClientHello without 0-RTT data.
...
 can increase the chance of making connections through those middleboxes
 by making the TLS 1.3 handshake look more like a TLS 1.2 handshake:
 
-- The client always provides a non-empty session ID in the ClientHello.
+- The client always provides a non-empty session ID in the ClientHello as
+  described in {{legacy_session_id}}.
 
 - If not offering early data, the client sends a dummy
-  change_cipher_spec record immediately before its second flight. This
+  change_cipher_spec record (see {{dummy_css}}
```
Comment (kaduk) [inline draft-ietf-tls-tls13.md]: (and here)

File: draft-ietf-tls-tls13.md [hunk 3]
```diff
@@ -5584,7 +5591,10 @@ TLS 1.2 session resumption, which improves the chance of successfully
 connecting through middleboxes. This "compatibility mode" is not explicitly
 negotiated. The client can opt to provide a session ID or not
 and the server has to echo it. Either side can send change_cipher_spec
-at any time during the handshake, as they must be ignored by the peer.
+at any time during the handshake, as it must be ignored by the peer.
+
+In the absence of out of band information, endpoints SHOULD use the
+techniques described here.
```
Comment (kaduk) [inline draft-ietf-tls-tls13.md]: For all time?

Comment (richsalz) [reply draft-ietf-tls-tls13.md]: Yes  Think about it.  If you know it's time to stop doing it, then you have out of band information.

File: draft-ietf-tls-tls13.md [hunk 4]
```diff
@@ -5560,17 +5563,21 @@ ClientHello without 0-RTT data.
...
 can increase the chance of making connections through those middleboxes
 by making the TLS 1.3 handshake look more like a TLS 1.2 handshake:
 
-- The client always provides a non-empty session ID in the ClientHello.
+- The client always provides a non-empty session ID in the ClientHello as
+  described in {{legacy_session_id}}.
 
 - If not offering early data, the client sends a dummy
-  change_cipher_spec record immediately before its second flight. This
+  change_cipher_spec record (see {{dummy_ccs}}
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: unbalanced paren

File: draft-ietf-tls-tls13.md [hunk 5]
```diff
@@ -1817,14 +1817,17 @@ random
   See {{implementation-notes}} for additional information.
 
 legacy_session_id
-: Versions of TLS before TLS 1.3 supported a "session resumption"
+: {#legacy_session_id} Versions
```
Comment (Lekensteyn) [inline draft-ietf-tls-tls13.md]: I don't think that you can use inline anchors like that, the parser does not seem to provide such a feature: https://github.com/cabo/kramdown-rfc2629

File: draft-ietf-tls-tls13.md [hunk 6]
```diff
@@ -5560,17 +5563,21 @@ ClientHello without 0-RTT data.
...
 and stable deployment of TLS 1.3 without 0-RTT prior to enabling 0-RTT.
 
-## Middlebox Compatibility Mode {#middlebox}
+## Network Intermediary Compatibility Mode {#intermediary}
 
-Field measurements have found that a significant number of middleboxes
-misbehave when a TLS client/server pair negotiates TLS 1.3. Implementations
+Field measurements have found that a significant number of network
+intermediaries, or "middleboxes,"
+can cause the TLS endpoints to fail to negotiate TLS 1.3.
```
Comment (Lekensteyn) [inline draft-ietf-tls-tls13.md]: Independent to this PR, but related: is it possible to provide a reference for these "field measurements"? Something like https://www.ietf.org/mail-archive/web/tls/current/msg24517.html (but preferably something more specific). Might be useful for the archeologists in the future.

Comment (richsalz) [reply draft-ietf-tls-tls13.md]: Perhaps bring that up on the list.  A non-normative link to @davidben's IETF-100 slides might be do-able.

**Comment by richsalz** [comment]
Pushed an updated commit that fixes the paren and anchor issue, thanks.

File: draft-ietf-tls-tls13.md [hunk 7]
```diff
@@ -3896,9 +3898,9 @@ TLS records are typed, which allows multiple higher-level protocols to
 be multiplexed over the same record layer. This document specifies
 four content types: handshake, application data, alert, and
 change_cipher_spec, with the latter being used only for
-compatibility purposes (see {{middlebox}}).
+compatibility purposes (see {{intermediary}}).
 
-An implementation may receive an unencrypted record of type
+{#dummy_ccs}An implementation may receive an unencrypted record of type
```
Comment (Lekensteyn) [inline draft-ietf-tls-tls13.md]: There is another unsupported inline anchor here. It seems also unhappy about the use of `client_hello`, maybe try `client-hello`?

Comment (richsalz) [reply draft-ietf-tls-tls13.md]: Thanks, I fixed this. And I moved a sentence from the 2nd to 3rd paragraph.

**Comment by martinthomson** [comment]
I'm having a little trouble with the justification for the use of "network intermediary" here.  The term "middlebox" has a well-established definition (see RFC 3234 for example).  This would seem to be less precise.

The use of "intermediary" implies something more than what we've observed.  I understand "intermediary" to involve protocol participation - TLS termination effectively - and that is not the sort of middlebox that is causing us issues, as David has explained on list.

Also, I realize that you think that the SHOULD is helpful.  I would prefer that our default mode of operation avoid the use of this hacky mode.  Firefox just shipped a pre-release with -22 and the compatibility hacks disabled and we can report an improvement in compatibility in at least one case, even without sending ChangeCipherSpec.  I can see both angles, but we might have to just disagree on this point.

**Comment by ekr** [comment]
On Wed, Dec 6, 2017 at 2:47 PM, Martin Thomson <notifications@github.com>
wrote
>
> Also, I realize that you think that the SHOULD is helpful. I would prefer
> that our default mode of operation avoid the use of this hacky mode.
> Firefox just shipped a pre-release with -22 and the compatibility hacks
> disabled and we can report an improvement in compatibility in at least one
> case, even without sending ChangeCipherSpec. I can see both angles, but we
> might have to just disagree on this point.
>
I go back and forth on the SHOULD, but it's something that would have to go
to the list...

**Comment by Lekensteyn** [comment]
> I would prefer that our default mode of operation avoid the use of this hacky mode. Firefox just shipped a pre-release with -22 and the compatibility hacks disabled and we can report an improvement in compatibility in at least one case.

Do you test both with and without CCS hack? Have you found cases where using CCS breaks while not sending CCS works?

FWIW, in tris I have chosen for unconditionally sending CCS to simplify the implementation (less knobs and code branches that needs to be tested).

**Comment by ekr** [comment]
On Wed, Dec 6, 2017 at 3:38 PM, Peter Wu <notifications@github.com> wrote:

> I would prefer that our default mode of operation avoid the use of this
> hacky mode. Firefox just shipped a pre-release with -22 and the
> compatibility hacks disabled and we can report an improvement in
> compatibility in at least one case.
>
> Do you test both with and without CCS hack?
>
We will be doing so shortly, but it's a 3-armed study (TLS 1.2, TLS 1.3-22,
TLS 1.3-compat), so we don't know how much we get from TLS-22 alone absent
historical controls.


> Have you found cases where using CCS breaks while not sending CCS works?
>
No, we do not have such cases.




FWIW, in tris I have chosen for unconditionally sending CCS to simplify the

**Comment by richsalz** [comment]
If folks don't like the intermediary term, so be it.  But it's important to collect all the compatibility changes into one place with appropriate links.

**Comment by ekr** [comment]
@richsalz: I added the links but omitted the name change and the SHOULD. At this point, a normative change like the SHOULD would need to go through the WG.

@richsalz: I see you aren't in acknowledgements. Please send me a PR to add yourself.

**Comment by richsalz** [comment]
Yes I am, right after Joe :)
The changes you made are okay.

**Pull request closed without merging**